File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/IA/ApreciacionFinalTratamientoService.java
package net.latinus.sistema.integral.gestion.seguridad.service.IA;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.ApreciacionFinalTratamientoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;

public interface ApreciacionFinalTratamientoService {
    
    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data ApreciacionFinalTratamientoDTO Paginada
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado PaginacionRequest datos para obtener todas las apreciaciones finales.
     * @param nemonicoMenu nemonico del menu.
     *
     * @return RespuestaPorDefectoAuditoria<PaginacionResponse<ApreciacionFinalTratamientoDTO>>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<ApreciacionFinalTratamientoDTO>> obtenerApreciacionesFinalesPaginado(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado, String nemonicoMenu);
    
    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data ApreciacionFinalTratamientoDTO si la apreciación final se creó/actualizó con éxito
     *
     * @param httpServletRequest request petición.
     * @param bodyEncriptado objeto ApreciacionFinalTratamientoDTO a crear/actualizar.
     * @param nemonicoMenu nemonico del menu.
     *
     * @return RespuestaPorDefectoAuditoria<ApreciacionFinalTratamientoDTO>
     */
    RespuestaPorDefectoAuditoria<ApreciacionFinalTratamientoDTO> crearApreciacionFinal(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado, String nemonicoMenu);
    
    /**
     * Elimina una apreciación final del sistema (elimina tanto situaciones como factores asociados)
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado ApreciacionFinalTratamientoDTO datos de la apreciación final a eliminar.
     *
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarApreciacionFinal(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
    
    /**
     * Elimina una situación específica del sistema
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado SituacionActualAdolescenteDTO datos de la situación a eliminar.
     *
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarSituacion(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
    
    /**
     * Elimina un factor específico del sistema
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado FactoresPresentesDTO datos del factor a eliminar.
     *
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarFactor(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
}

File: pronacej-front/src/app/modules/seguridad/services/apreciacionFinalTratamiento.service.ts
import { Injectable } from '@angular/core';
import { ApreciacionFinalTratamientoDTO } from 'app/core/model/both/apreciacionFinalTratamientoDTO.model';
import { BodyEncriptado } from 'app/core/model/both/bodyEncriptado.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable, Subscriber } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class ApreciacionFinalTratamientoService {
    private path = '/apreciacion-final-tratamiento';

    constructor(private backendService: BackendService) {}

    /**
     * Obtiene las apreciaciones finales del tratamiento disponibles de manera paginada
     *
     * @param paginacionRequest PaginacionRequest objeto con los parámetros de paginación
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<PaginacionResponse<ApreciacionFinalTratamientoDTO>>>
     */
    obtenerApreciacionesFinalesPaginado(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu = ''
    ): Observable<
        RespuestaPorDefecto<PaginacionResponse<ApreciacionFinalTratamientoDTO>>
    > {
        let endPoint = this.path + '/obtenerApreciacionesFinalesPaginado';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    /**
     * Crea una apreciación final del tratamiento con los datos enviados en el request
     *
     * @param apreciacionFinalDTO ApreciacionFinalTratamientoDTO datos de la apreciación final a crear
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<ApreciacionFinalTratamientoDTO>>
     */
    crearApreciacionFinal(
        apreciacionFinalDTO: ApreciacionFinalTratamientoDTO,
        nemonicoMenu: string = ''
    ): Observable<RespuestaPorDefecto<ApreciacionFinalTratamientoDTO>> {
        let endPoint = this.path + '/crearApreciacionFinal';
        return this.backendService.postFinal(
            endPoint,
            apreciacionFinalDTO,
            nemonicoMenu
        );
    }

    /**
     * Elimina una apreciación final del tratamiento
     *
     * @param apreciacionFinalDTO ApreciacionFinalTratamientoDTO datos de la apreciación final a eliminar
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<boolean>>
     */
    eliminarApreciacionFinal(
        apreciacionFinalDTO: ApreciacionFinalTratamientoDTO,
        nemonicoMenu = ''
    ):Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/eliminarApreciacionFinal';

        return this.backendService.postFinal(
            endPoint,
            apreciacionFinalDTO,
            nemonicoMenu
        );
    }

    /**
     * Verifica y maneja los errores generales del servicio
     *
     * @param error any error a verificar
     * @param mostrarError boolean indica si se debe mostrar el error
     * @returns any
     */
    async checkError(error: any, mostrarError = true):Promise<string> {
        return await this.backendService.checkError(error, mostrarError);
    }
}
